File: frontend/clean.js
const fs = require('fs');
let js = fs.readFileSync('src/main.js', 'utf8');

// 4a
let start = js.indexOf('// ── DEDICATED ADMIN TAB CONTROLLER');
let endStr = "alert('✔ Đã thêm (local test)');\r\n    }\r\n  }\r\n}";
let end = js.indexOf(endStr, start);
if(end === -1) end = js.indexOf("alert('✔ Đã thêm (local test)');\n    }\n  }\n}", start);
if(end !== -1) {
    js = js.substring(0, start) + js.substring(end + endStr.length);
} else {
    // try shorter
    let endStr2 = "alert('✔ Đã thêm (local test)');\r\n  }\r\n}";
    let end2 = js.indexOf(endStr2, start);
    if(end2 !== -1) {
        js = js.substring(0, start) + js.substring(end2 + endStr2.length);
    }
}

// 4b
js = js.replace(/checkAdminAuth\(\);\r?\n/, '');

// 4c
let s4c = "if(document.getElementById('page-admin').classList.contains('show')){\r\n    renderAdminLocList();\r\n  }";
if(js.includes(s4c)) js = js.replace(s4c, '');
else js = js.replace("if(document.getElementById('page-admin').classList.contains('show')){\n    renderAdminLocList();\n  }", '');

// 4d
js = js.replace(/\['home','map','cook','admin'\]\.forEach/g, "['home','map','cook'].forEach");
js = js.replace(/if\(tab==='admin'\) renderAdminTab\(\);\r?\n/, "");

let p1 = js.indexOf("if(tab==='admin'){ aiFab.style.opacity='0'; aiFab.style.pointerEvents='none'; aiFab.style.zIndex='-1'; }");
let p2 = js.indexOf("else{ aiFab.style.opacity=''; aiFab.style.pointerEvents=''; aiFab.style.zIndex=''; }", p1);
if(p1 !== -1 && p2 !== -1) {
    js = js.substring(0, p1) + "aiFab.style.opacity=''; aiFab.style.pointerEvents=''; aiFab.style.zIndex='';\n" + js.substring(p2 + 86);
}

// 4e
js = js.replace(/\['a-cat','edit-cat','s-cat'\]\.forEach/g, "['s-cat'].forEach");

// 4f
let fIdx1 = js.indexOf("['ar-cat','er-cat'].forEach(function(selId){");
if (fIdx1 !== -1) {
    let fIdx2 = js.indexOf("});", fIdx1);
    js = js.substring(0, fIdx1) + js.substring(fIdx2 + 3);
}

let rcpPillsStr = "var adminRcpPills = document.getElementById('admin-rcp-filter-pills');";
let rIdx1 = js.indexOf(rcpPillsStr);
if (rIdx1 !== -1) {
    let rIdx2 = js.indexOf("adminRcpPills.innerHTML = h;", rIdx1);
    if (rIdx2 !== -1) {
        js = js.substring(0, rIdx1) + js.substring(rIdx2 + 28);
    }
}

// 4g
js = js.replace(/\['page-home','page-cook','page-admin'\]\.forEach/g, "['page-home','page-cook'].forEach");

// 4h
js = js.replace(/\s*'addLocation':\s*\{[^}]*\},\r?\n/g, "\n");
js = js.replace(/\s*'updateLocation':\s*\{[^}]*\},\r?\n/g, "\n");
js = js.replace(/\s*'deleteLocation':\s*\{[^}]*\},\r?\n/g, "\n");
js = js.replace(/\s*'setGeminiAPIKey':\s*\{[^}]*\},\r?\n/g, "\n");
js = js.replace(/\s*'getAdminStatus':\s*\{[^}]*\},\r?\n/g, "\n");

let mStart = js.indexOf("if (gasMethodName === 'addLocation') {");
if (mStart !== -1) {
    let mEnd = js.indexOf("else if (gasMethodName === 'saveSuggestion') {", mStart);
    if (mEnd !== -1) {
        js = js.substring(0, mStart) + "if (gasMethodName === 'saveSuggestion') {" + js.substring(mEnd + 46);
    }
}

// 4i
const win_vars = [
    'changeAdminLocPage', 'changeAdminRcpPage', 'deleteAdminLoc', 'deleteRecipeItem',
    'doAdminAdd', 'doAdminAddRecipe', 'doAdminUpdateRecipe', 'openAdminEdit',
    'openAdminEditRecipe', 'openAdminModal', 'saveAdminEdit', 'setAdminLocFilter',
    'setAdminRcpFilter', 'switchAdminSec'
];
for(const v of win_vars) {
    js = js.replace(new RegExp('window\\.' + v + '\\s*=\\s*' + v + ';\\s*', 'g'), '');
}

fs.writeFileSync('src/main.js', js, 'utf8');
console.log('Done!');

File: frontend/clean.py
import codecs
with codecs.open('src/main.js', 'r', 'utf-8') as f:
    js = f.read()

# 4a.
start_str = '// ── DEDICATED ADMIN TAB CONTROLLER'
start_idx = js.find(start_str)
if start_idx != -1:
    end_str = 'function doAdminAdd()'
    end_idx = js.find(end_str, start_idx)
    # find the end of doAdminAdd
    next_func_idx = js.find('\n//', end_idx) # next section usually starts with //
    if next_func_idx == -1: next_func_idx = js.find('\nfunction', end_idx+100)
    
    js = js[:start_idx] + js[next_func_idx:]

# 4b.
js = js.replace('checkAdminAuth();', '')

# 4c.
js = js.replace("if(document.getElementById('page-admin').classList.contains('show')){\n    renderAdminLocList();\n  }", "")
js = js.replace("if(document.getElementById('page-admin').classList.contains('show')){\r\n    renderAdminLocList();\r\n  }", "")
# If one liner
import re
js = re.sub(r"if\(document\.getElementById\('page-admin'\)\.classList\.contains\('show'\)\)\{\s*renderAdminLocList\(\);\s*\}", "", js)

# 4d.
js = js.replace("['home','map','cook','admin'].forEach", "['home','map','cook'].forEach")
js = re.sub(r"if\(tab==='admin'\)\s*renderAdminTab\(\);", "", js)
js = re.sub(r"if\(tab==='admin'\)\{\s*aiFab\.style\.opacity='0';\s*aiFab\.style\.pointerEvents='none';\s*aiFab\.style\.zIndex='-1';\s*\}\s*else\{\s*(aiFab\.style\.opacity='';[^}]+)\}", r"\1", js)

# 4e.
js = js.replace("['a-cat','edit-cat','s-cat'].forEach", "['s-cat'].forEach")

# 4f.
js = re.sub(r"\['ar-cat','er-cat'\]\.forEach\([\s\S]*?\}\);", "", js)
js = re.sub(r"var\s*adminRcpPills\s*=\s*document\.getElementById\('admin-rcp-filter-pills'\);[\s\S]*?adminRcpPills\.innerHTML\s*=\s*h;", "", js)

# 4g.
js = js.replace("['page-home','page-cook','page-admin'].forEach", "['page-home','page-cook'].forEach")

# 4h.
js = re.sub(r"\s*'addLocation':\s*\{[^\}]*\},", "", js)
js = re.sub(r"\s*'updateLocation':\s*\{[^\}]*\},", "", js)
js = re.sub(r"\s*'deleteLocation':\s*\{[^\}]*\},", "", js)
js = re.sub(r"\s*'setGeminiAPIKey':\s*\{[^\}]*\},", "", js)
js = re.sub(r"\s*'getAdminStatus':\s*\{[^\}]*\},", "", js)

js = re.sub(r"if\s*\(gasMethodName\s*===\s*'addLocation'\)\s*\{[\s\S]*?\}\s*else\s*if\s*\(gasMethodName\s*===\s*'updateLocation'\)\s*\{[\s\S]*?\}\s*else\s*if\s*\(gasMethodName\s*===\s*'deleteLocation'\)\s*\{[\s\S]*?\}\s*else\s*if\s*\(gasMethodName\s*===\s*'setGeminiAPIKey'\)\s*\{[\s\S]*?\}\s*else\s*if\s*\(gasMethodName\s*===\s*'saveSuggestion'\)", "if (gasMethodName === 'saveSuggestion')", js)

# 4i.
win_vars = [
    'changeAdminLocPage', 'changeAdminRcpPage', 'deleteAdminLoc', 'deleteRecipeItem',
    'doAdminAdd', 'doAdminAddRecipe', 'doAdminUpdateRecipe', 'openAdminEdit',
    'openAdminEditRecipe', 'openAdminModal', 'saveAdminEdit', 'setAdminLocFilter',
    'setAdminRcpFilter', 'switchAdminSec'
]
for v in win_vars:
    js = re.sub(r"window\." + v + r"\s*=\s*" + v + r";\s*", "", js)

with codecs.open('src/main.js', 'w', 'utf-8') as f:
    f.write(js)
print('cleaned')
